File: n8n_data-ingestion-node.py
"""
Data Ingestion Service for n8n Trading Processor
Permanent version with direct node access for reliable data retrieval.
"""

import pandas as pd
from datetime import datetime, timedelta
from typing import Dict, List, Tuple, Any
import warnings

# Suppress warnings for cleaner output
warnings.filterwarnings('ignore')

def last_trading_date(today=None):
    """Return last trading date (Mon–Fri), mapping Sat/Sun -> Fri."""
    dt = pd.Timestamp(today or datetime.now())
    if dt.weekday() == 5:  # Sat -> Fri
        return (dt - pd.Timedelta(days=1)).normalize()
    if dt.weekday() == 6:  # Sun -> Fri
        return (dt - pd.Timedelta(days=2)).normalize()
    return dt.normalize()

def process_data_ingestion():
    """Main data ingestion function for n8n with direct node access."""
    try:
        # Initialize data containers
        google_sheets_data = []
        ticker_data = []
        
        # Get all market data from the merged input
        all_market_data = []
        input_items = _input.all()
        
        # Deduplicate market data by symbol to prevent processing duplicates
        seen_symbols = set()
        for item in input_items:
            if 'data' in item['json'] and isinstance(item['json']['data'], dict):
                symbol = item['json']['data'].get('symbol', '').upper()
                if symbol and symbol not in seen_symbols:
                    all_market_data.append(item['json'])
                    seen_symbols.add(symbol)
        
        # Define benchmark and S&P 500 tickers
        benchmarks = {'IWO', 'XBI', 'SPY', 'IWM'}
        sp500_tickers = {'^GSPC', 'SPY', '^SPX'}
        
        # 1. Get Google Sheets portfolio data directly
        try:
            google_sheets_raw = _('Google Sheets Node').all()
            for item in google_sheets_raw:
                if hasattr(item['json'], 'to_py'):
                    python_data = item['json'].to_py()
                else:
                    python_data = dict(item['json'])
                
                # Validate Google Sheets data structure
                if all(key in python_data for key in ['Date', 'Ticker', 'Shares', 'Buy Price']):
                    google_sheets_data.append(python_data)
        except Exception as e:
            # Log error but continue processing
            pass

        # 2. Get ticker list from Parse Current Holdings
        try:
            ticker_raw = _('Parse Current Holdings').all()
            for item in ticker_raw:
                if hasattr(item['json'], 'to_py'):
                    python_data = item['json'].to_py()
                else:
                    python_data = dict(item['json'])
                
                if 'ticker' in python_data:
                    ticker_data.append(python_data)
        except Exception as e:
            # Log error but continue processing
            pass
        
        # Initialize classified data structure
        classified_data = {
            'historical_portfolio': google_sheets_data,
            'current_market_data': [],
            'benchmark_data': [],
            'sp500_data': [],
            'ticker_list': ticker_data
        }
        
        # Process all market data from the merged input and classify by ticker type
        for item in all_market_data:
            if 'data' in item and isinstance(item['data'], dict):
                market_item = item['data']
                ticker = market_item.get('symbol', '').upper()
                
                # Skip invalid tickers
                if not ticker or ticker in ['BY-TYPE', 'UNKNOWN']:
                    continue
                
                # Standardize market data using DXFeed fields
                standardized_item = {
                    'Ticker': ticker,
                    'Date': market_item.get('summary-date', datetime.now().strftime('%Y-%m-%d')),
                    # Real OHLC data from DXFeed
                    'Open': float(market_item.get('open', market_item.get('prev-close', 0))),
                    'High': float(market_item.get('day-high-price', market_item.get('last', 0))),
                    'Low': float(market_item.get('day-low-price', market_item.get('last', 0))),
                    'Close': float(market_item.get('close', market_item.get('last', 0))),
                    'Volume': 0,  # Not provided in this format
                    'PrevClose': float(market_item.get('prev-close', 0)),
                    'HasIntradayData': True,
                    'DataSource': 'TastyTrade_DXFeed_Direct_Access',
                    # Additional market data
                    'Bid': float(market_item.get('bid', 0)),
                    'Ask': float(market_item.get('ask', 0)),
                    'Mid': float(market_item.get('mid', 0)),
                    'Mark': float(market_item.get('mark', 0)),
                    'YearLow': float(market_item.get('year-low-price', 0)),
                    'YearHigh': float(market_item.get('year-high-price', 0)),
                    'InstrumentType': market_item.get('instrument-type', ''),
                    'UpdatedAt': market_item.get('updated-at', ''),
                    'IsTradingHalted': market_item.get('is-trading-halted', False)
                }
                
                # Classify by ticker type
                if ticker in sp500_tickers:
                    classified_data['sp500_data'].append(standardized_item)
                elif ticker in benchmarks:
                    classified_data['benchmark_data'].append(standardized_item)
                else:
                    classified_data['current_market_data'].append(standardized_item)
        
        # Enhanced debug info
        debug_info = {
            "processing_date": datetime.now().isoformat(),
            "data_access_method": "direct_node_access",
            "data_sources": {
                "google_sheets_available": len(google_sheets_data) > 0,
                "ticker_list_available": len(ticker_data) > 0,
                "total_market_data_items": len(all_market_data),
                "input_items_received": len(input_items),
                "unique_symbols_processed": len(seen_symbols)
            },
            "classified_counts": {
                "historical_portfolio": len(classified_data['historical_portfolio']),
                "current_market_data": len(classified_data['current_market_data']),
                "benchmark_data": len(classified_data['benchmark_data']),
                "sp500_data": len(classified_data['sp500_data']),
                "ticker_list": len(classified_data['ticker_list'])
            },
            "data_quality": {
                "all_tickers_have_intraday": True,
                "data_source": "TastyTrade_DXFeed_Direct_Access_Permanent"
            },
            "tickers_found": {
                "portfolio": [item['Ticker'] for item in classified_data['current_market_data']],
                "benchmarks": [item['Ticker'] for item in classified_data['benchmark_data']],
                "sp500": [item['Ticker'] for item in classified_data['sp500_data']]
            }
        }
        
        # Return classified data in n8n format - single output regardless of input count
        return [{
            "json": {
                "type": "classified_data",
                **classified_data,
                "debug_info": debug_info
            }
        }]
        
    except Exception as e:
        import traceback
        return [{
            "json": {
                "type": "error",
                "error": f"Data ingestion failed: {str(e)}",
                "traceback": traceback.format_exc(),
                "timestamp": datetime.now().isoformat()
            }
        }]

# Execute the main function
return process_data_ingestion()

File: n8n_data-ingestion-node.js
// Data Ingestion Service for n8n Trading Processor
// Uses named node access to fetch data from previous nodes and handles multiple data formats.

/**
 * Returns the last trading date (Mon–Fri), mapping Sat/Sun -> Fri.
 * @param {Date | null} today - The date to start from. Defaults to current date if null.
 * @returns {Date} The last trading date.
 */
function lastTradingDate(today = null) {
    let dt = today ? new Date(today) : new Date();
    dt.setHours(0, 0, 0, 0); // Normalize to start of day

    const dayOfWeek = dt.getDay(); // 0 is Sunday, 1 is Monday, ..., 6 is Saturday

    if (dayOfWeek === 6) { // Saturday
        dt.setDate(dt.getDate() - 1); // Friday
    } else if (dayOfWeek === 0) { // Sunday
        dt.setDate(dt.getDate() - 2); // Friday
    }
    return dt;
}

/**
 * Main data ingestion function for n8n with named node access.
 * @returns {Promise<any[]>} An array of n8n items containing classified data or an error.
 */
return (async function() {
    try {
        // --- Initialize data containers ---
        let currentHoldingsData = [];
        let portfolioMarketData = [];
        let benchmarkMarketData = [];
        let allMarketData = [];
        let fetchedCashBalance = 0.0;

        // --- Fetch data explicitly from each source node by name ---
        const holdingsItems = $('Parse Current Holdings').all();
        currentHoldingsData = holdingsItems.map(i => i.json);
        const portfolioTickers = new Set(currentHoldingsData.map(h => h.symbol.toUpperCase()));


        const portfolioMarketItems = $('Get Portfolio Asset Snapshots').all();
        const parsedPortfolioData = portfolioMarketItems
            .filter(item => item.json)
            .flatMap(item => {
                if (item.json.snapshots) {
                    // Handle crypto format, which has a 'snapshots' key
                    const snapshots = item.json.snapshots;
                    return Object.keys(snapshots).map(ticker => {
                        const standardizedTicker = ticker.replace('/', ''); // Convert 'BTC/USD' to 'BTCUSD'
                        return { ...snapshots[ticker], symbol: standardizedTicker };
                    });
                } else if (Object.keys(item.json).length > 0) {
                    // Handle stock format, which has the ticker as the key
                    const ticker = Object.keys(item.json)[0];
                    const data = item.json[ticker];
                    if (data) { // Ensure data exists for the ticker
                        return { ...data, symbol: ticker };
                    }
                }
                return []; // Return empty array for items that don't match known formats
            });
        portfolioMarketData.push(...parsedPortfolioData);

        const benchmarkMarketItems = $('Market Data Benchmark Summary Node').all();
        const parsedBenchmarkData = benchmarkMarketItems
            .filter(item => item.json)
            .flatMap(item => {
                return Object.keys(item.json).map(ticker => ({
                    ...item.json[ticker],
                    symbol: ticker
                }));
            });
        benchmarkMarketData.push(...parsedBenchmarkData);

        const accountSnapshotItems = $('Get Account Snapshot').all();
        if (accountSnapshotItems.length > 0 && accountSnapshotItems[0].json && accountSnapshotItems[0].json.cash) {
            fetchedCashBalance = parseFloat(accountSnapshotItems[0].json.cash);
        }

        // --- Combine all data sources ---
        allMarketData.push(...portfolioMarketData, ...benchmarkMarketData);

        // --- Process and classify all combined market data ---
        const seenSymbols = new Set();
        const uniqueMarketData = allMarketData.filter(item => {
            const symbol = item.symbol || '';
            if (!symbol || seenSymbols.has(symbol)) {
                return false;
            }
            seenSymbols.add(symbol);
            return true;
        });

        const sp500Tickers = new Set(['^GSPC', 'SPY', '^SPX']);

        let classifiedData = {
            historical_portfolio: currentHoldingsData,
            current_market_data: [],
            benchmark_data: [],
            sp500_data: [],
            current_cash_balance: fetchedCashBalance,
        };

        for (const marketItem of uniqueMarketData) {
            const ticker = (marketItem.symbol || '').toUpperCase();
            if (!ticker || ['BY-TYPE', 'UNKNOWN'].includes(ticker)) {
                continue;
            }
            
            const standardizedItem = {
                Ticker: ticker,
                Date: marketItem.dailyBar?.t ? new Date(marketItem.dailyBar.t).toISOString().slice(0, 10) : new Date().toISOString().slice(0, 10),
                Open: parseFloat(marketItem.dailyBar?.o || 0),
                High: parseFloat(marketItem.dailyBar?.h || 0),
                Low: parseFloat(marketItem.dailyBar?.l || 0),
                Close: parseFloat(marketItem.dailyBar?.c || marketItem.latestTrade?.p || 0),
                Volume: parseFloat(marketItem.dailyBar?.v || 0),
                PrevClose: parseFloat(marketItem.prevDailyBar?.c || 0),
                DataSource: 'Alpaca_Snapshot_API',
            };
            
            if (sp500Tickers.has(ticker)) {
                classifiedData.sp500_data.push(standardizedItem);
            } else if (portfolioTickers.has(ticker)) {
                classifiedData.current_market_data.push(standardizedItem);
            } else {
                classifiedData.benchmark_data.push(standardizedItem);
            }
        }
        
        // --- Return classified data in n8n format ---
        return [{
            json: {
                type: "classified_data",
                ...classifiedData
            }
        }];
        
    } catch (e) {
        console.error(`Data ingestion failed: ${e.message}`);
        return [{
            json: {
                type: "error",
                error: `Data ingestion failed: ${e.message}`,
                traceback: e.stack,
                timestamp: new Date().toISOString()
            }
        }];
    }
})();
